### PR TITLE
Fix/mobile sidebar revamp

### PR DESCRIPTION
**Pull request by nikrb**
supercedes pr 132 fix/mobile-sidebar.

Major changes to sidebar so we don't lose tree state after hiding the material-ui drawer (mobile only).
Slimmed down CMSContainer component.
Html element details/summary didn't work as a controlled component so replace with home grown version.
Move Collapsible test to utils (as that's where they belong) and create a new Collapsible specific snapshot test.
Create tests for CMSContainer and add tests to Sidebar and Sidebar utils.

### CODE REVIEW

File: frontend/src/Sidebar/Collapsible.test.js [hunk 1]
```diff
@@ -1,94 +1,31 @@
...
-//   </BrowserRouter>
-// );
+it('should expand on click', () => {
+  const expand = jest.fn();
+  const comp = shallow(
+    <Collapsible id="" title="test title" open={false} expanded={expand}>
+      {{ _id: '1', title: 'Some article title' }}
+    </Collapsible>
+  );
+  comp.instance().onClick({ stopPropagation: () => {}});
```
Comment (jenovs) [inline frontend/src/Sidebar/Collapsible.test.js]: `Prettier` re-formats this line.

File: frontend/src/Sidebar/Sidebar.js [hunk 2]
```diff
@@ -1,51 +1,141 @@
...
+    let { mobile } = this.state;
+    const windowWidth = window.innerWidth;
+
+    if (mobile && windowWidth > BREAK_MOBILE) {
+      mobile = false;
+    } else if (!mobile && windowWidth <= BREAK_MOBILE) {
+      mobile = true;
+    }
+    return mobile;
+  };
```
Comment (jenovs) [inline frontend/src/Sidebar/Sidebar.js]: Could be refactored out of component and get called with parameters:
```js
const checkMobile = (mobile, windowWidth) => {
  const BREAK_MOBILE = 900;

  if (
    (mobile && windowWidth > BREAK_MOBILE) ||
    (!mobile && windowWidth <= BREAK_MOBILE)
  ) {
    return !mobile;
  }

  return mobile;
};
```

File: frontend/src/Sidebar/Sidebar.test.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    },
+    Introduction: '103',
+  },
+};
+
+const match = {
+  params: { id: '' },
+};
+
+function flushPromises() {
```
Comment (jenovs) [inline frontend/src/Sidebar/Sidebar.test.js]: Linter is not happy :unamused:

File: frontend/src/Sidebar/Sidebar.test.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+it('should show mobile drawer', () => {
+  expect.hasAssertions();
+  comp.setState({ articlesHtml: null, mobile: true });
+  inst.openDrawer();
+  return Promise.resolve().then(() => {
+    comp.update();
+    expect(inst.state.open).toBe(true);
+    expect(comp).toMatchSnapshot();
+  });
+});
```
Comment (jenovs) [inline frontend/src/Sidebar/Sidebar.test.js:110]: Passing of this test depends on the previous test. Not good.
Edit: actually looks like all of the test are depending on each other :unamused:

File: frontend/src/Sidebar/utils.test.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+it('should create html tree from articles', () => {
+  const articleTree = getTree(articles);
+  const html = buildHtml(articles, articleTree, '', onArticleSelect, onExpand);
+  expect(html).toMatchSnapshot();
+});
+
+it('should create expanded tree if id present', () => {
+  const articleTree = getTree(articles);
+  const html = buildHtml(articles, articleTree, '101', onArticleSelect, onExpand);
```
Comment (jenovs) [inline frontend/src/Sidebar/utils.test.js]: `Prettier` re-formats this line.

**Review by jenovs — APPROVED** [review]

File: frontend/src/Sidebar/Sidebar.test.js (hunk 4 above)
Comment (nikrb) [reply frontend/src/Sidebar/Sidebar.test.js:110]: It seemed natural for one test to run on the results of the last - I take it this is bad practice. Will fix

Comment (nikrb) [reply frontend/src/Sidebar/Sidebar.test.js:110]: fixed with beforeEach but doesn't seem to have been caught (this review still shows for me)

Comment (jenovs) [reply frontend/src/Sidebar/Sidebar.test.js:110]: EDIT: I don't think it's natural for one test to run on the results of the last.
Better way to accomplish that would be:
```js
it('should test some stuff', () => {
  // initial setup
  // assert that setup is correct
  // simulate user interaction, e.g. click
  // assert changes
  // simulate another interaction
  // assert changes
  // etc.
}
```

**Review by jenovs — APPROVED** [review]

File: frontend/src/Sidebar/Sidebar.test.js (hunk 4 above)
Comment (nikrb) [reply frontend/src/Sidebar/Sidebar.test.js:110]: @jenovs is this comment out of order? you're saying tests following on isn't natural ... or do you mean the beforeEach isn't the best way to fix?

Comment (nikrb) [reply frontend/src/Sidebar/Sidebar.test.js:110]: oh I see you mean put them in the same test - ok. Will merge shortly

Comment (jenovs) [reply frontend/src/Sidebar/Sidebar.test.js:110]: Looks like we wrote our comments almost simultaneously :grinning: 
I edited my comment.
EDIT: Looks like GitHub comments are not quite real time.

**Pull request merged**